x/bank/keeper.go

```diff
@@ -258,3 +297,56 @@ func inputOutputCoins(ctx sdk.Context, am auth.AccountKeeper, inputs []Input, ou
 
 	return allTags, nil
 }
+
+func delegateCoins(
+	ctx sdk.Context, ak auth.AccountKeeper, addr sdk.AccAddress, amt sdk.Coins,
+) (sdk.Tags, sdk.Error) {
+
+	ctx.GasMeter().ConsumeGas(costSubtractCoins, "delegateCoins")
+
+	oldCoins := getCoins(ctx, ak, addr)
+	newCoins := oldCoins.Minus(amt)
+
+	if !newCoins.IsNotNegative() {
+		return nil, sdk.ErrInsufficientCoins(fmt.Sprintf("%s < %s", oldCoins, amt))
+	}
+
+	va, ok := ak.GetAccount(ctx, addr).(auth.VestingAccount)
+	if ok {
+		blockTime := ctx.BlockHeader().Time
+
+		va.TrackDelegation(blockTime, amt)
+		ak.SetAccount(ctx, va)
+	} else {
+		if err := setCoins(ctx, ak, addr, newCoins); err != nil {
+			return nil, err
+		}
+	}
+
+	return sdk.NewTags("sender", []byte(addr.String())), nil
```

Comment:
shouldn't we add an `action` here as well ?

Reply:
Good question. Is this the standard now?

Reply:
I guess so... Maybe @cwgoes or @sunnya97 will know better?

Reply:
Done.